### DISCUSSION TITLE
OpenAPI generation: Listing scopes for Non-OAuth2 schemes per OpenAPI 3.1

### DISCUSSION THREAD

**Discussion by whiskeyriver** [fastapi/fastapi#6713] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from typing import List

from fastapi import FastAPI, Request, Security
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer, SecurityScopes
from pydantic import BaseModel


class CurrentUser(BaseModel):
    class Config:
        arbitrary_types_allowed = True

    credentials: HTTPAuthorizationCredentials
    required_scopes: List[str]

def Authentication(
    request: Request,
    scopes: SecurityScopes,
    auth: HTTPAuthorizationCredentials = Security(HTTPBearer())
) -> CurrentUser:
    user = CurrentUser(credentials=auth, required_scopes=scopes.scopes)
    request.state.current_user = user
    return user


app = FastAPI(debug=True)

@app.get("/checkit", response_model=CurrentUser)
def checkit(current_user: CurrentUser = Security(Authentication, scopes=["checkit:read", "checkit:update"])):
    return current_user
```

### Description

While my Authentication function correctly receives the scopes for the route, they will not be reported in `openapi.json`.

Resulting `openapi.json`:
```json
    "paths": {
        "/checkit": {
            "get": {
                "security": [
                    {
                        "HTTPBearer": []
                    }
                ]
            }
        }
    },
```

### Description: Reporting `SecurityScopes` in HTTPBearer array

OpenAPI 3.1 (https://spec.openapis.org/oas/v3.1.0#patterned-fields-2) allows roles/scopes to be specified in the scopes array of security schemes other than OAuth.

I.e.
```json
    "paths": {
        "/with_scope": {
            "get": {
                "security": [
                    {
                        "HTTPBearer": ["entity:read", "entity:create"]
                    }
                ]
            }
        }
    },
```

### The solution you would like

It wasn't clear to me how to customize the openapi generation to support this. Advice on implementing this for my project, or a standardized approach, would be very much appreciated.

### Environment

* FastAPI Version: 0.66.0
* Python Version: 3.8

**Comment by whiskeyriver** [comment]
I've made some headway on this.

I see that the provided scopes are only passed on to the `SecurityRequirement` if the dependency type is `OpenIdConnect` or `OAuth2`: https://github.com/tiangolo/fastapi/blob/master/fastapi/dependencies/utils.py#L146

I would like to propose that the `SecurityBase` model get an attibute, like `use_scopes`, that is by default False. https://github.com/tiangolo/fastapi/blob/master/fastapi/security/base.py#L4. From there, each subclass could specify whether it wants to enable this, with `OpenIdConnect` and `OAuth2` defaulting to True, and `HTTPBase` defaulting to False.

Then `get_sub_dependant` would check the `use_scopes` attribute, rather than `isinstance(dependency, (OAuth2, OpenIdConnect))`.

This way, all I have to do to enable this in my own code is:
```python
def Authentication(
    request: Request,
    scopes: SecurityScopes,
    auth: HTTPAuthorizationCredentials = Security(HTTPBearer(use_scopes=True))
) -> CurrentUser:
    user = CurrentUser(credentials=auth, required_scopes=scopes.scopes)
    request.state.current_user = user
    return user
```

and here's my openapi.json:
```json
        "/checkit": {
            "get": {
                "security": [
                    {
                        "HTTPBearer": [
                            "checkit:read",
                            "checkit:update"
                        ]
                    }
                ]
            }
        }
```

I don't think `use_scopes` is the right name for the attribute to pass to `HTTPBearer`, I just used it as a placeholder to mimic the variable in `get_sub_dependant`.

Here's a PR I opened in my own fork to better display what I mean: https://github.com/whiskeyriver/fastapi/pull/1/files

**Comment by agropper** [comment]
This relates to my issue https://github.com/tiangolo/fastapi/issues/3344#issuecomment-860040098 where, as a rank novice, I'm trying to use FastAPI to model work-in-progress on IETF GNAP:

https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-06.html and
https://www.ietf.org/archive/id/draft-ietf-gnap-resource-servers-01.html

When it comes to scopes, the OAuth 2 practice has been a problem and is now being revised to OAuth RAR https://datatracker.ietf.org/doc/draft-ietf-oauth-rar/ 

GNAP is not backward compatible with OAuth 2 for other reasons but shares the RAR upgrade See 4. in https://www.ietf.org/archive/id/draft-ietf-gnap-core-protocol-06.html#appendix-B

**Comment by whiskeyriver** [comment, marked as answer]
When circling back on this I noticed the `openapi_extra` parameter documented at https://fastapi.tiangolo.com/advanced/path-operation-advanced-configuration/#openapi-extra

For what I'm trying to do this is a perfectly fine solution.

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue 👍

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
